src/native-desktop/RootView.tsx

```diff
@@ -0,0 +1,165 @@
+//   React Native doesn't offer a convenient mechanism to peek into keyboard/mouse events globally (the way
+// addEventListener achieves in the Web counterpart), so we rely on a spying view we render.
+// 2. Redirection to Input.ts
+// 3. A place for the root FocusManager
+type Constructor<T extends React.Component> = new (...args: any[]) => T;
+
+function applyDesktopBehaviorMixin<TRootViewBase extends Constructor<React.Component>>(RootViewBase: TRootViewBase) {
+    return class RootView extends RootViewBase implements React.ChildContextProvider<any> {
+        static childContextTypes: React.ValidationMap<any> = {
+            focusManager: PropTypes.object
+        };
+
+        _focusManager: FocusManager;
+        _keyboardHandlerInstalled = false;
+        _isNavigatingWithKeyboard: boolean = false;
+        _isNavigatingWithKeyboardUpateTimer: number | undefined;
+
+        constructor(...args: any[]) {
+            super(...args);
+            // Initialize the root FocusManager which is aware of all
+            // focusable elements.
+            this._focusManager = new FocusManager(undefined);
+        }
+
+        _onTouchStartCapture = (e: SyntheticEvent) => {
+            this._updateKeyboardNavigationState(false);
+        }
+
+        _onKeyDownCapture = (e: SyntheticEvent) => {
+
+            let kbdEvent = EventHelpers.toKeyboardEvent(e);
+            if (kbdEvent.keyCode === KEY_CODE_TAB) {
+                this._updateKeyboardNavigationState(true);
+            }
+
+            if (kbdEvent.keyCode === KEY_CODE_ESC) {
+                // If Esc is pressed and the focused element stays the same after some time,
+                // switch the keyboard navigation off to dismiss the outline.
+                const activeComponent = FocusManager.getCurrentFocusedComponent();
+
+                if (this._isNavigatingWithKeyboardUpateTimer) {
+                    clearTimeout(this._isNavigatingWithKeyboardUpateTimer);
+                }
+
+                this._isNavigatingWithKeyboardUpateTimer = setTimeout(() => {
+                    this._isNavigatingWithKeyboardUpateTimer = undefined;
+
+                    if (activeComponent === FocusManager.getCurrentFocusedComponent()) {
+                        this._updateKeyboardNavigationState(false);
+                    }
+                }, 500);
+            }
+        }
+
+        _updateKeyboardNavigationState(isNavigatingWithKeyboard: boolean) {
+            if (this._isNavigatingWithKeyboardUpateTimer) {
+                window.clearTimeout(this._isNavigatingWithKeyboardUpateTimer);
+                this._isNavigatingWithKeyboardUpateTimer = undefined;
+            }
+
+            if (this._isNavigatingWithKeyboard !== isNavigatingWithKeyboard) {
+                this._isNavigatingWithKeyboard = isNavigatingWithKeyboard;
+
+                UserInterface.keyboardNavigationEvent.fire(isNavigatingWithKeyboard);
+            }
```

Comment:
>} [](start = 12, length = 1)

How does the focus outline get rendered? It looks like the web version supports different focus outlines depending on whether the navigation was done with keyboard.

Reply:
We just defer to the UWP system focus rectangles (black rectangles) for now. Its displaying basically mimics the flag we have for "keyboard navigation on"", with some exceptions (the flag gets reset when pressing ESC, whereas I haven't found a way to convey this to UWP as well).
It's true web implementation offers 2 outlines. Also S4L disables that completely and uses an internal "focus outline" extension that shows a red rectangle. I chose not to cover that for now until it's clear why we have all these random initiatives and what we actually want.

---
In reply to: [159357484](https://github.com/Microsoft/reactxp/pull/404#discussion_r159357484) [](ancestors = 159357484)